dind/Dockerfile

```diff
@@ -1,26 +1,35 @@
 ARG DOCKER_VERSION
 
-FROM docker:${DOCKER_VERSION}-dind-rootless
+FROM golang:alpine AS build
+WORKDIR /build
+COPY docker-iptables-glue/go.* ./
+RUN go mod download
+COPY docker-iptables-glue/ ./
+RUN CGO_ENABLED=0 GOOS=linux go build
 
+FROM docker:${DOCKER_VERSION}-dind-rootless
 USER root
 
-RUN export CRUN_LATEST=$(wget -qO- https://api.github.com/repos/containers/crun/releases/latest | sed -En 's/.*"tag_name"\s*:\s*"((\\"|[^"])*)".*/\1/p') && \
-    wget -qO /usr/local/bin/crun "https://github.com/containers/crun/releases/download/${CRUN_LATEST}/crun-${CRUN_LATEST}-linux-amd64" && \
-    chmod +x /usr/local/bin/crun && \
-    export SLIRP4NETNS_LATEST=$(wget -qO- https://api.github.com/repos/rootless-containers/slirp4netns/releases/latest | sed -En 's/.*"tag_name"\s*:\s*"((\\"|[^"])*)".*/\1/p') && \
-    wget -qO /usr/local/bin/slirp4netns "https://github.com/rootless-containers/slirp4netns/releases/download/${SLIRP4NETNS_LATEST}/slirp4netns-x86_64" && \
-    chmod +x /usr/local/bin/slirp4netns && \
-    mkdir /opt/containerd && \
-    chown rootless:rootless /opt/containerd && \
-    mkdir /run/netns && \
-    chown rootless:rootless /run/netns
+ENV CRUN_BASE_DOWNLOAD_PATH=https://github.com/containers/crun/releases/download \
+    SLIRP4NETNS_BASE_DOWNLOAD_PATH=https://github.com/rootless-containers/slirp4netns/releases/download
 
-COPY ./init /sbin/init
+RUN apk add --no-cache iproute2 socat jq && \
+    export CRUN_LATEST=$(wget -qO- https://api.github.com/repos/containers/crun/releases/latest | jq -r .tag_name) && \
+    wget -qO /usr/local/bin/crun "${CRUN_BASE_DOWNLOAD_PATH}/${CRUN_LATEST}/crun-${CRUN_LATEST}-linux-amd64" && \
+    export SLIRP4NETNS_LATEST=$(wget -qO- https://api.github.com/repos/rootless-containers/slirp4netns/releases/latest | jq -r .tag_name) && \
+    wget -qO /usr/local/bin/slirp4netns "${SLIRP4NETNS_BASE_DOWNLOAD_PATH}/${SLIRP4NETNS_LATEST}/slirp4netns-x86_64" && \
+    chmod +x /usr/local/bin/crun /usr/local/bin/slirp4netns && \
+    mkdir /opt/containerd /run/netns && \
+    chown rootless:rootless /opt/containerd /run/netns
 
-USER rootless
+COPY ./init /sbin/
+COPY --from=build /build/docker-iptables-glue /sbin/
+RUN chmod u+s /sbin/docker-iptables-glue
 
-ENV DOCKERD_ROOTLESS_ROOTLESSKIT_NET slirp4netns
-ENV DOCKERD_ROOTLESS_ROOTLESSKIT_MTU 65520
-ENV DOCKERD_ENTRYPOINT_ARGS --add-runtime crun=/usr/local/bin/crun --default-runtime crun --experimental --registry-mirror https://mirror.gcr.io
+USER rootless
+ENV DOCKER_HOST=tcp://localhost:2375 \
+    DOCKERD_ROOTLESS_ROOTLESSKIT_NET=slirp4netns \
+    DOCKERD_ROOTLESS_ROOTLESSKIT_MTU=65520 \
+    DOCKERD_ENTRYPOINT_ARGS="--add-runtime crun=/usr/local/bin/crun --default-runtime crun --experimental --registry-mirror https://mirror.gcr.io"
```

Comment:
ENV instructions don't create additional Docker layers, so we could revert this to the previous state and just add:

```
ENV DOCKER_HOST tcp://localhost:2375
ENV DOCKER_TLS_CERTDIR
```

I've added `DOCKER_TLS_CERTDIR` so it is consistent with the use of the port `2375`. Otherwise, it would be `2376` by default.

Reply:
aeab21d